### PR TITLE
PSF parser

### PR DESCRIPTION
**Pull request by stefdoerr**
I wrote a PSF parser based off the PDB parser. PSF does not include much info so it's much simpler.
Essentially only atom info and bonds are of use to you. The rest are impropers and dihedrals which won't be used for visualization.

PSF is space delimited so split works (not like column formats i.e. PDB).

I don't know how to test it though. Any tips?

### CODE REVIEW

**Comment by stefdoerr** [comment]
Hm actually it might not be space delimited as I thought (although in most cases it practically becomes so). Will need to work more on this.

File: src/parser/psf-parser.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+                        if( from < to ){
+                            ap1.index = from;
+                            ap2.index = to;
+                        }else{
+                            ap1.index = to;
+                            ap2.index = from;
+                        }
+
+                        // interpret records where a bond is given multiple times as double/triple bonds
```
Comment (arose) [inline src/parser/psf-parser.js]: is this used in psf files

File: src/parser/psf-parser.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        atomStore.resize( Math.round( this.streamer.data.length / 80 ) );
+
+        var ap1 = s.getAtomProxy();
+        var ap2 = s.getAtomProxy();
+
+        var idx = 0;
+        var modelIdx = 0;
+        var pendingStart = true;
+
+        var residins_regex = /(\d+)([a-zA-Z])/;
```
Comment (arose) [inline src/parser/psf-parser.js]: camelCase please, pythonista :-)

File: src/parser/psf-parser.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            }
+        }
+
+        this.streamer.eachChunkOfLines( function( lines/*, chunkNo, chunkCount*/ ){
+            _parseChunkOfLines( 0, lines.length, lines );
+        } );
+
+        sb.finalize();
+
+        if( unitcellDict.a !== undefined ){
```
Comment (arose) [inline src/parser/psf-parser.js]: there is no unitcell data, right? However, if there is data about the simulation box, that can be put here, for example as in the gro parser

```javascript
s.unitcell = new Unitcell(
  boxes[ 0 ][ 0 ], boxes[ 0 ][ 4 ], boxes[ 0 ][ 8 ],
  90, 90, 90, "P 1"
);
```

File: src/parser/psf-parser.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                        inscode = '';
+                    }
+                    resname = atominfo[3].trim();
+                    atomname = atominfo[4].trim();
+                    element = atominfo[5].trim();
+
+                    atomStore.growIfFull();
+                    atomStore.atomTypeId[idx] = atomMap.add(atomname, element);
+                    atomStore.serial[idx] = serial;
+                    atomStore.bfactor[idx] = 0;
```
Comment (arose) [inline src/parser/psf-parser.js]: no need to assign bfactor, altloc, accupancy

File: src/parser/psf-parser.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+ * @private
+ */
+
+
+import { Matrix4 } from "../../lib/three.es6.js";
+
+import { Debug, Log, ParserRegistry } from "../globals.js";
+import StructureParser from "./structure-parser.js";
+import Unitcell from "../symmetry/unitcell.js";
+import Assembly from "../symmetry/assembly.js";
```
Comment (arose) [inline src/parser/psf-parser.js]: can be removed

**Comment by arose** [comment]
Cool! Thanks for doing this.

> I don't know how to test it though. Any tips?

`npm test` will run a linter (eslint) and then the unittests (with mocha). The linter will catch lots of syntax/best practice issues. On how to test the parser you can have a look at `test/parser/tests-pdb-parser.js`. You can run a single test file with `mocha test/parser/tests-pdb-parser.js`.

I also like to add an example for the webapp (`examples/js/examples.js`) for new features.

**Comment by stefdoerr** [comment]
Great, thanks! Sorry for the poor pull request. I will take a look at it this week hopefully :)

**Comment by arose** [comment]
Hey, I made a few changes to how parsers are handled. Have a look at the [changes](https://github.com/arose/ngl/commit/2d184d2403630041241638d32826e9a911e1e8b5#diff-469ae544b78a4b14e77acc00aeda98dd) to the gro parser. Also, to use the parser in the build it is now sufficient to import it in [ngl.js](https://github.com/arose/ngl/blob/master/src/ngl.js#L645) with `import PsfParser from "./parser/psf-parser.js";`.

**Comment by arose** [comment]
Hey @stefdoerr any plans for finishing this? Would be quite cool!

**Comment by stefdoerr** [comment]
Sorry for the late answer @arose. Since I wrote the HTMDTrajectory class I didn't need ngl to read PSF files anymore so I forgot about this.
The problem is that my JS and JS debugging skills are close to none so it's too much of an effort right now for me. However the reader ought to be quite simple and you are free to re-implement it from my python code: https://github.com/Acellera/htmd/blob/master/htmd/molecule/readers.py#L609

Practically just one `if-elseif` to determine which section (called `mode` in my code) you are in and another one to parse the lines of that section using string splitting.

**Comment by arose** [comment]
ok, added it in e7e0ddbffd3ab983b1d92c707d9eebc5c57a72ee

**Pull request closed without merging**